README.md
# @jcamp/tailwindcss-plugin-animate

<div align="center">
  <img src="./.github/tailwindcss-mark.svg" alt="Tailwind CSS" width="108" height="66">
  <h1>Tailwind CSS Animation Plugin</h1>
  <p>An updated animation plugin for Tailwind CSS v3</p>
</div>

Easily add animations to your Tailwind projects. Uses the animations from [animate.css](https://github.com/animate-css/animate.css) but through Tailwind's system.

```html
<div class="animate-shake-x animate-delay animate-infinite">(icon)</div>
```

Keyframes are only added to the generated CSS if an animation using them is added.

## Install

```bash
npm i -D @jcamp/tailwindcss-plugin-animate
```

Then add the plugin to your `tailwind.config.js` file:

```js
// tailwind.config.js
const animate = require('@jcamp/tailwindcss-plugin-animate')
module.exports = {
  theme: {
    // ...
  },
  plugins: [
    animate({
      /* options */
    }),
    // ...
  ],
}
```

```ts
// tailwind.config.ts
import type { Config } from 'tailwindcss'
import animate from '@jcamp/tailwindcss-plugin-animate'

export default <Partial<Config>>{
  theme: {
    // ...
  },
  plugins: [
    animate({
      /* options */
    }),
  ],
}
```

## Configuration

Refer to the [type definition](https://github.com/jcamp-code/tailwindcss-plugin-animate/blob/main/src/types.ts) for all configurations avaliable.

## Credits

This plugin is heaviliy inspired by and based on the work of [tailwindcss-animate.css](https://github.com/bentzibentz/tailwindcss-animate.css) created by [Fabian Bentz](https://github.com/bentzibentz). Unfortunately, this package has not been updated in a few years and not for Tailwind v3 either.

It is also includes some ideas by [Jamie Kyle](https://github.com/jamiebuilds) and his work at [tailwindcss-animate](https://github.com/jamiebuilds/tailwindcss-animate). I wanted more keyframes built in, but liked his ideas for matching utilities.

## License

MIT License &copy; 2022-PRESENT [John Campion](https://github.com/JohnCampionJr/)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## v0.2.2

[compare changes](https://github.com/jcamp-code/tailwindcss-plugin-animate/compare/v0.2.0...v0.2.1)

### 🩹 Fixes

- Remove postinstall script ([a9538c9](https://github.com/jcamp-code/tailwindcss-plugin-animate/commit/a9538c9))

## v0.2.0

[compare changes](https://github.com/jcamp-code/tailwindcss-plugin-animate/compare/v0.1.0...v0.2.0)

### 🚀 Enhancements

BREAKING CHANGE: May require tweaks to tailwind config file, see README for details.

- Updated to latest tailwind plugin structure ([d54d4a6](https://github.com/jcamp-code/tailwindcss-plugin-animate/commit/d54d4a6))

## [0.1.0] - 2022-08-13

Initial release!


tests/index.test.js
const path = require('path')
const animate = require('../dist/index.js')
const postcss = require('postcss')
const tailwindcss = require('tailwindcss')

function run(config, css = '@tailwind utilities', plugin = tailwindcss) {
  let { currentTestName } = expect.getState()
  config = {
    ...{
      plugins: [animate(config.animate)],
      corePlugins: {
        preflight: false,
      },
    },
    ...config,
  }

  return postcss(plugin(config)).process(css, {
    from: `${path.resolve(__filename)}?test=${currentTestName}`,
  })
}

it('hasDuration1000', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-duration-1000"></div>` }],
  }

  return run(config, '@tailwind utilities').then(result => {
    const value = String.raw`.animate-duration-1000 {
    animation-duration: 1000ms
}`
    expect(result.css).toContain(value)
  })
})

it('hasDuration2000', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-duration-2000"></div>` }],
  }

  return run(config, '@tailwind utilities').then(result => {
    const value = String.raw`.animate-duration-2000 {
    animation-duration: 2000ms
}`
    expect(result.css).toContain(value)
  })
})

it('hasDurationArbitrary', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-duration-[2500]"></div>` }],
  }

  return run(config, '@tailwind utilities').then(result => {
    const value = String.raw`.animate-duration-\[2500\] {
    animation-duration: 2500ms
}`
    expect(result.css).toContain(value)
  })
})

it('hasAnimateShake-X', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-shake-x"></div>` }],
  }

  return run(config, '@tailwind utilities').then(result => {
    const value = String.raw`.animate-shake-x {
    animation-name: shake-x;
    animation-duration: 1000ms
}
@keyframes shake-x {`

    expect(result.css).toContain(value)
  })
})

it('hasAnimateShakeX', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-shakeX"></div>` }],
  }

  return run(config, '@tailwind utilities').then(result => {
    const value = String.raw`.animate-shakeX {
    animation-name: shakeX;
    animation-duration: 1000ms
}
@keyframes shakeX {`

    expect(result.css).toContain(value)
  })
})

it('warnsOnArbitrary', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-[nonsense]"></div>` }],
  }
  console.warn = jest.fn()

  return run(config, '@tailwind utilities').then(result => {
    expect(console.warn).toHaveBeenCalledTimes(1)
  })
})

it('includesRunning', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-running"></div>` }],
  }

  return run(config, '@tailwind utilities').then(result => {
    const value = String.raw`.animate-running {
    animation-play-state: running
}`

    expect(result.css).toContain(value)
  })
})

it('includesExtraPropsForHeartBeat', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-heart-beat"></div>` }],
  }

  return run(config, '@tailwind utilities').then(result => {
    const value = String.raw`.animate-heart-beat {
    animation-name: heart-beat;
    animation-duration: 1300ms;
    animation-timing-function: ease-in-out
}
@keyframes heart-beat {`

    expect(result.css).toContain(value)
  })
})

it('includesExtraPropsForHeartBeatWithConfig', () => {
  const config = {
    animate: { heartBeatSpeed: 2000 },
    content: [{ raw: String.raw`<div class="animate-heart-beat"></div>` }],
  }

  return run(config, '@tailwind utilities').then(result => {
    const value = String.raw`.animate-heart-beat {
    animation-name: heart-beat;
    animation-duration: 2000ms;
    animation-timing-function: ease-in-out
}
@keyframes heart-beat {`

    expect(result.css).toContain(value)
  })
})

it('usesPrefixOption', () => {
  const config = {
    animate: { prefix: 'an-' },
    content: [{ raw: String.raw`<div class="an-shakeX"></div>` }],
  }

  return run(config, '@tailwind utilities').then(result => {
    const value = String.raw`.an-shakeX {
    animation-name: shakeX;
    animation-duration: 1000ms
}
@keyframes shakeX {`

    expect(result.css).toContain(value)
  })
})


tests/index.test.ts
import path from 'node:path'
import postcss from 'postcss'
import tailwindcss from 'tailwindcss'
import animate from '../dist/index.js'

function run(config, css = '@tailwind utilities', plugin = tailwindcss) {
  const { currentTestName } = expect.getState()
  config = {
    ...{
      plugins: [animate(config.animate)],
      corePlugins: {
        preflight: false,
      },
    },
    ...config,
  }

  return postcss(plugin(config)).process(css, {
    from: `${path.resolve(__filename)}?test=${currentTestName}`,
  })
}

it('hasDuration1000', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-duration-1000"></div>` }],
  }

  return run(config, '@tailwind utilities').then((result) => {
    const value = String.raw`.animate-duration-1000 {
    animation-duration: 1000ms
}`
    expect(result.css).toContain(value)
  })
})

it('hasDuration2000', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-duration-2000"></div>` }],
  }

  return run(config, '@tailwind utilities').then((result) => {
    const value = String.raw`.animate-duration-2000 {
    animation-duration: 2000ms
}`
    expect(result.css).toContain(value)
  })
})

it('hasDurationArbitrary', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-duration-[2500]"></div>` }],
  }

  return run(config, '@tailwind utilities').then((result) => {
    const value = String.raw`.animate-duration-\[2500\] {
    animation-duration: 2500ms
}`
    expect(result.css).toContain(value)
  })
})

it('hasAnimateShake-X', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-shake-x"></div>` }],
  }

  return run(config, '@tailwind utilities').then((result) => {
    const value = String.raw`.animate-shake-x {
    animation-name: shake-x;
    animation-duration: 1000ms
}
@keyframes shake-x {`

    expect(result.css).toContain(value)
  })
})

it('hasAnimateShakeX', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-shakeX"></div>` }],
  }

  return run(config, '@tailwind utilities').then((result) => {
    const value = String.raw`.animate-shakeX {
    animation-name: shakeX;
    animation-duration: 1000ms
}
@keyframes shakeX {`

    expect(result.css).toContain(value)
  })
})

it('warnsOnArbitrary', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-[nonsense]"></div>` }],
  }
  console.warn = jest.fn()

  return run(config, '@tailwind utilities').then((result) => {
    expect(console.warn).toHaveBeenCalledTimes(1)
  })
})

it('includesRunning', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-running"></div>` }],
  }

  return run(config, '@tailwind utilities').then((result) => {
    const value = String.raw`.animate-running {
    animation-play-state: running
}`

    expect(result.css).toContain(value)
  })
})

it('includesExtraPropsForHeartBeat', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-heart-beat"></div>` }],
  }

  return run(config, '@tailwind utilities').then((result) => {
    const value = String.raw`.animate-heart-beat {
    animation-name: heart-beat;
    animation-duration: 1300ms;
    animation-timing-function: ease-in-out
}
@keyframes heart-beat {`

    expect(result.css).toContain(value)
  })
})

it('includesExtraPropsForHeartBeatWithConfig', () => {
  const config = {
    animate: { heartBeatSpeed: 2000 },
    content: [{ raw: String.raw`<div class="animate-heart-beat"></div>` }],
  }

  return run(config, '@tailwind utilities').then((result) => {
    const value = String.raw`.animate-heart-beat {
    animation-name: heart-beat;
    animation-duration: 2000ms;
    animation-timing-function: ease-in-out
}
@keyframes heart-beat {`

    expect(result.css).toContain(value)
  })
})

it('usesPrefixOption', () => {
  const config = {
    animate: { prefix: 'an-' },
    content: [{ raw: String.raw`<div class="an-shakeX"></div>` }],
  }

  return run(config, '@tailwind utilities').then((result) => {
    const value = String.raw`.an-shakeX {
    animation-name: shakeX;
    animation-duration: 1000ms
}
@keyframes shakeX {`

    expect(result.css).toContain(value)
  })
})


tests/animate.test.js
const path = require('path')
const animate = require('../dist/animate.js')
const postcss = require('postcss')
const tailwindcss = require('tailwindcss')

function run(config, css = '@tailwind utilities', plugin = tailwindcss) {
  let { currentTestName } = expect.getState()
  config = {
    ...{
      plugins: [animate(config.animate)],
      corePlugins: {
        preflight: false,
      },
    },
    ...config,
  }

  return postcss(plugin(config)).process(css, {
    from: `${path.resolve(__filename)}?test=${currentTestName}`,
  })
}

it('hasDuration1000', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-duration-1000"></div>` }],
  }

  return run(config, '@tailwind utilities').then(result => {
    const value = String.raw`.animate-duration-1000 {
    animation-duration: 1000ms
}`
    expect(result.css).toContain(value)
  })
})

it('hasDuration2000', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-duration-2000"></div>` }],
  }

  return run(config, '@tailwind utilities').then(result => {
    const value = String.raw`.animate-duration-2000 {
    animation-duration: 2000ms
}`
    expect(result.css).toContain(value)
  })
})

it('hasDurationArbitrary', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-duration-[2500]"></div>` }],
  }

  return run(config, '@tailwind utilities').then(result => {
    const value = String.raw`.animate-duration-\[2500\] {
    animation-duration: 2500ms
}`
    expect(result.css).toContain(value)
  })
})

it('hasAnimateShake-X', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-shake-x"></div>` }],
  }

  return run(config, '@tailwind utilities').then(result => {
    const value = String.raw`.animate-shake-x {
    animation-name: shake-x;
    animation-duration: 1000ms
}
@keyframes shake-x {`

    expect(result.css).toContain(value)
  })
})

it('hasAnimateShakeX', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-shakeX"></div>` }],
  }

  return run(config, '@tailwind utilities').then(result => {
    const value = String.raw`.animate-shakeX {
    animation-name: shakeX;
    animation-duration: 1000ms
}
@keyframes shakeX {`

    expect(result.css).toContain(value)
  })
})

it('warnsOnArbitrary', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-[nonsense]"></div>` }],
  }
  console.warn = jest.fn()

  return run(config, '@tailwind utilities').then(result => {
    expect(console.warn).toHaveBeenCalledTimes(1)
  })
})

it('includesRunning', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-running"></div>` }],
  }

  return run(config, '@tailwind utilities').then(result => {
    const value = String.raw`.animate-running {
    animation-play-state: running
}`

    expect(result.css).toContain(value)
  })
})

it('includesExtraPropsForHeartBeat', () => {
  const config = {
    content: [{ raw: String.raw`<div class="animate-heart-beat"></div>` }],
  }

  return run(config, '@tailwind utilities').then(result => {
    const value = String.raw`.animate-heart-beat {
    animation-name: heart-beat;
    animation-duration: 1300ms;
    animation-timing-function: ease-in-out
}
@keyframes heart-beat {`

    expect(result.css).toContain(value)
  })
})

it('includesExtraPropsForHeartBeatWithConfig', () => {
  const config = {
    animate: { heartBeatSpeed: 2000 },
    content: [{ raw: String.raw`<div class="animate-heart-beat"></div>` }],
  }

  return run(config, '@tailwind utilities').then(result => {
    const value = String.raw`.animate-heart-beat {
    animation-name: heart-beat;
    animation-duration: 2000ms;
    animation-timing-function: ease-in-out
}
@keyframes heart-beat {`

    expect(result.css).toContain(value)
  })
})


src/types.ts
export interface AnimateOptions {
  /**
   * Class prefix for matching animation rules.
   *
   * @default `animate-`
   */
  prefix?: string

  /**
   * Default animation speed / duration
   *
   * @default 1000
   */
  animatedSpeed?: number

  /**
   * Default heart beat animation speed / duration
   *
   * @default 1300
   */
  heartBeatSpeed?: number

  /**
   * Default hinge animation speed / duration
   *
   * @default 2000
   */
  hingeSpeed?: number

  /**
   * Default bounce in animation speed / duration
   *
   * @default 750
   */
  bounceInSpeed?: number

  /**
   * Default bounce out animation speed / duration
   *
   * @default 750
   */
  bounceOutSpeed?: number

  /**
   * Default animay delay for animate-delay utility
   *
   * @default 1000
   */
  animationDelaySpeed?: number
}


jest/custom-matchers.js
const prettier = require('prettier')
const { diff } = require('jest-diff')

function format(input) {
  return prettier.format(input, {
    parser: 'css',
    printWidth: 100,
  })
}

expect.extend({
  // Compare two CSS strings with all whitespace removed
  // This is probably naive but it's fast and works well enough.
  toMatchCss(received, argument) {
    function stripped(str) {
      return str.replace(/\s/g, '').replace(/;/g, '')
    }

    const options = {
      comment: 'stripped(received) === stripped(argument)',
      isNot: this.isNot,
      promise: this.promise,
    }

    const pass = stripped(received) === stripped(argument)

    const message = pass
      ? () => {
          return (
            this.utils.matcherHint(
              'toMatchCss',
              undefined,
              undefined,
              options
            ) +
            '\n\n' +
            `Expected: not ${this.utils.printExpected(format(received))}\n` +
            `Received: ${this.utils.printReceived(format(argument))}`
          )
        }
      : () => {
          const actual = format(received)
          const expected = format(argument)

          const diffString = diff(expected, actual, {
            expand: this.expand,
          })

          return (
            this.utils.matcherHint(
              'toMatchCss',
              undefined,
              undefined,
              options
            ) +
            '\n\n' +
            (diffString && diffString.includes('- Expect')
              ? `Difference:\n\n${diffString}`
              : `Expected: ${this.utils.printExpected(expected)}\n` +
                `Received: ${this.utils.printReceived(actual)}`)
          )
        }

    return { actual: received, message, pass }
  },
  toIncludeCss(received, argument) {
    function stripped(str) {
      return str
        .replace('/* prettier-ignore */', '')
        .replace(/\s/g, '')
        .replace(/;/g, '')
    }

    const options = {
      comment: 'stripped(received).includes(stripped(argument))',
      isNot: this.isNot,
      promise: this.promise,
    }

    const pass = stripped(received).includes(stripped(argument))

    const message = pass
      ? () => {
          return (
            this.utils.matcherHint(
              'toIncludeCss',
              undefined,
              undefined,
              options
            ) +
            '\n\n' +
            `Expected: not ${this.utils.printExpected(format(received))}\n` +
            `Received: ${this.utils.printReceived(format(argument))}`
          )
        }
      : () => {
          const actual = format(received)
          const expected = format(argument)

          const diffString = diff(expected, actual, {
            expand: this.expand,
          })

          return (
            this.utils.matcherHint(
              'toIncludeCss',
              undefined,
              undefined,
              options
            ) +
            '\n\n' +
            (diffString && diffString.includes('- Expect')
              ? `Difference:\n\n${diffString}`
              : `Expected: ${this.utils.printExpected(expected)}\n` +
                `Received: ${this.utils.printReceived(actual)}`)
          )
        }

    return { actual: received, message, pass }
  },
})

expect.extend({
  // Compare two CSS strings with all whitespace removed
  // This is probably naive but it's fast and works well enough.
  toMatchFormattedCss(received = '', argument = '') {
    function format(input) {
      return prettier.format(input.replace(/\n/g, ''), {
        parser: 'css',
        printWidth: 100,
      })
    }
    const options = {
      comment: 'stripped(received) === stripped(argument)',
      isNot: this.isNot,
      promise: this.promise,
    }

    let formattedReceived = format(received)
    let formattedArgument = format(argument)

    const pass = formattedReceived === formattedArgument

    const message = pass
      ? () => {
          return (
            this.utils.matcherHint(
              'toMatchCss',
              undefined,
              undefined,
              options
            ) +
            '\n\n' +
            `Expected: not ${this.utils.printExpected(formattedReceived)}\n` +
            `Received: ${this.utils.printReceived(formattedArgument)}`
          )
        }
      : () => {
          const actual = formattedReceived
          const expected = formattedArgument

          const diffString = diff(expected, actual, {
            expand: this.expand,
          })

          return (
            this.utils.matcherHint(
              'toMatchCss',
              undefined,
              undefined,
              options
            ) +
            '\n\n' +
            (diffString && diffString.includes('- Expect')
              ? `Difference:\n\n${diffString}`
              : `Expected: ${this.utils.printExpected(expected)}\n` +
                `Received: ${this.utils.printReceived(actual)}`)
          )
        }

    return { actual: received, message, pass }
  },
})


src/keyframes/keyframes.ts
const keyframeBounce = {
  'from, 20%, 53%, 80%, to': {
    animationTimingFunction: 'cubic-bezier(0.215, 0.61, 0.355, 1)',
    transform: 'translate3d(0, 0, 0)',
  },
  '40%, 43%': {
    animationTimingFunction: 'cubic-bezier(0.755, 0.05, 0.855, 0.06)',
    transform: 'translate3d(0, -30px, 0)',
  },
  '70%': {
    animationTimingFunction: 'cubic-bezier(0.755, 0.05, 0.855, 0.06)',
    transform: 'translate3d(0, -15px, 0)',
  },
  '90%': {
    transform: 'translate3d(0, -4px, 0)',
  },
}

const keyframeFlash = {
  'from, 50%, to': {
    opacity: '1',
  },
  '25%, 75%': {
    opacity: '0',
  },
}

const keyframePulse = {
  from: {
    transform: 'scale3d(1, 1, 1)',
  },
  '50%': {
    transform: 'scale3d(1.05, 1.05, 1.05)',
  },
  to: {
    transform: 'scale3d(1, 1, 1)',
  },
}

const keyframeRubberBand = {
  from: {
    transform: 'scale3d(1, 1, 1)',
  },
  '30%': {
    transform: 'scale3d(1.25, 0.75, 1)',
  },
  '40%': {
    transform: 'scale3d(0.75, 1.25, 1)',
  },
  '50%': {
    transform: 'scale3d(1.15, 0.85, 1)',
  },
  '65%': {
    transform: 'scale3d(0.95, 1.05, 1)',
  },
  '75%': {
    transform: 'scale3d(1.05, 0.95, 1)',
  },
  to: {
    transform: 'scale3d(1, 1, 1)',
  },
}

const keyframeShakeX = {
  'from, to': {
    transform: 'translate3d(0, 0, 0)',
  },
  '10%, 30%, 50%, 70%, 90%': {
    transform: 'translate3d(-10px, 0, 0)',
  },
  '20%, 40%, 60%, 80%': {
    transform: 'translate3d(10px, 0, 0)',
  },
}

const keyframeShakeY = {
  'from, to': {
    transform: 'translate3d(0, 0, 0)',
  },
  '10%, 30%, 50%, 70%, 90%': {
    transform: 'translate3d(0, -10px, 0)',
  },
  '20%, 40%, 60%, 80%': {
    transform: 'translate3d(0, 10px, 0)',
  },
}

const keyframeHeadShake = {
  '0%': {
    transform: 'translateX(0)',
  },
  '6.5%': {
    transform: 'translateX(-6px) rotateY(-9deg)',
  },
  '18.5%': {
    transform: 'translateX(5px) rotateY(7deg)',
  },
  '31.5%': {
    transform: 'translateX(-3px) rotateY(-5deg)',
  },
  '43.5%': {
    transform: 'translateX(2px) rotateY(3deg)',
  },
  '50%': {
    transform: 'translateX(0)',
  },
}

const keyframeSwing = {
  '20%': {
    transform: 'rotate3d(0, 0, 1, 15deg)',
  },
  '40%': {
    transform: 'rotate3d(0, 0, 1, -10deg)',
  },
  '60%': {
    transform: 'rotate3d(0, 0, 1, 5deg)',
  },
  '80%': {
    transform: 'rotate3d(0, 0, 1, -5deg)',
  },
  to: {
    transform: 'rotate3d(0, 0, 1, 0deg)',
  },
}

const keyframeTada = {
  from: {
    transform: 'scale3d(1, 1, 1)',
  },
  '10%, 20%': {
    transform: 'scale3d(0.9, 0.9, 0.9) rotate3d(0, 0, 1, -3deg)',
  },
  '30%, 50%, 70%, 90%': {
    transform: 'scale3d(1.1, 1.1, 1.1) rotate3d(0, 0, 1, 3deg)',
  },
  '40%, 60%, 80%': {
    transform: 'scale3d(1.1, 1.1, 1.1) rotate3d(0, 0, 1, -3deg)',
  },
  to: {
    transform: 'scale3d(1, 1, 1)',
  },
}

const keyframeWobble = {
  from: {
    transform: 'translate3d(0, 0, 0)',
  },
  '15%': {
    transform: 'translate3d(-25%, 0, 0) rotate3d(0, 0, 1, -5deg)',
  },
  '30%': {
    transform: 'translate3d(20%, 0, 0) rotate3d(0, 0, 1, 3deg)',
  },
  '45%': {
    transform: 'translate3d(-15%, 0, 0) rotate3d(0, 0, 1, -3deg)',
  },
  '60%': {
    transform: 'translate3d(10%, 0, 0) rotate3d(0, 0, 1, 2deg)',
  },
  '75%': {
    transform: 'translate3d(-5%, 0, 0) rotate3d(0, 0, 1, -1deg)',
  },
  to: {
    transform: 'translate3d(0, 0, 0)',
  },
}

const keyframeJello = {
  'from, 11.1% to': {
    transform: 'translate3d(0, 0, 0)',
  },
  '22.2%': {
    transform: 'skewX(-12.5deg) skewY(-12.5deg)',
  },
  '33.3%': {
    transform: 'skewX(6.25deg) skewY(6.25deg)',
  },
  '44.4%': {
    transform: 'skewX(-3.125deg) skewY(-3.125deg)',
  },
  '55.5%': {
    transform: 'skewX(1.5625deg) skewY(1.5625deg)',
  },
  '66.6%': {
    transform: 'skewX(-0.78125deg) skewY(-0.78125deg)',
  },
  '77.7%': {
    transform: 'skewX(0.390625deg) skewY(0.390625deg)',
  },
  '88.8%': {
    transform: 'skewX(-0.1953125deg) skewY(-0.1953125deg)',
  },
}

const keyframeHeartBeat = {
  '0%': {
    transform: 'scale(1)',
  },
  '14%': {
    transform: 'scale(1.3)',
  },
  '28%': {
    transform: 'scale(1)',
  },
  '42%': {
    transform: 'scale(1.3)',
  },
  '70%': {
    transform: 'scale(1)',
  },
}

const keyframeHinge = {
  '0%': {
    transformOrigin: 'top left',
    animationTimingFunction: 'ease-in-out',
  },
  '20%, 60%': {
    transform: 'rotate3d(0, 0, 1, 80deg)',
    transformOrigin: 'top left',
    animationTimingFunction: 'ease-in-out',
  },
  '40%, 80%': {
    transform: 'rotate3d(0, 0, 1, 60deg)',
    transformOrigin: 'top left',
    animationTimingFunction: 'ease-in-out',
  },
  to: {
    transform: 'translate3d(0, 700px, 0)',
    opacity: '0',
  },
}

const keyframeJackInTheBox = {
  from: {
    opacity: '0',
    transformOrigin: 'center bottom',
    transform: 'scale(0.1) rotate(30deg)',
  },
  '50%': {
    transform: 'rotate(-10deg)',
  },
  '70%': {
    transform: 'rotate(3deg)',
  },
  to: {
    transform: 'scale(1)',
  },
}

exports.keyframeBounce = keyframeBounce
exports.keyframeFlash = keyframeFlash
exports.keyframePulse = keyframePulse
exports.keyframeRubberBand = keyframeRubberBand
exports.keyframeShakeX = keyframeShakeX
exports.keyframeShakeY = keyframeShakeY
exports.keyframeHeadShake = keyframeHeadShake
exports.keyframeSwing = keyframeSwing
exports.keyframeTada = keyframeTada
exports.keyframeWobble = keyframeWobble
exports.keyframeJello = keyframeJello
exports.keyframeHeartBeat = keyframeHeartBeat
exports.keyframeHinge = keyframeHinge
exports.keyframeJackInTheBox = keyframeJackInTheBox


src/index.ts
import plugin from 'tailwindcss/plugin'
import { camelCase, kebabCase } from 'lodash'
import type { PluginCreator } from 'tailwindcss/types/config'
import type { AnimateOptions } from './types'
import { availableKeyframes } from './keyframes'

function createPluginAnimate(options: AnimateOptions = {}): PluginCreator {
  const PREFIX = options.prefix ? options.prefix : 'animate-'

  function createAnimateValues() {
    const values = {}
    const keys = Object.keys(availableKeyframes)

    keys.forEach((el) => {
      const animationName = el.replace('@keyframes ', '')
      const className = kebabCase(animationName)

      // lets both animate-shakeX and animate-shake-x work
      values[className] = className
      values[animationName] = animationName
    })

    return values
  }

  function processValue(value: unknown): string {
    if (isNaN(Number(value))) {
      return value as string
    }
    return `${value}ms`
  }

  // set fallback if speed not defined
  const heartBeatSpeed = options.heartBeatSpeed ? options.heartBeatSpeed : 1300
  const hingeSpeed = options.hingeSpeed ? options.hingeSpeed : 2000
  const bounceInSpeed = options.bounceInSpeed ? options.bounceInSpeed : 750
  const bounceOutSpeed = options.bounceOutSpeed ? options.bounceOutSpeed : 750
  const animationDelaySpeed = options.animationDelaySpeed
    ? options.animationDelaySpeed
    : 500

  const extraProperties = {
    bounce: {
      transformOrigin: 'center bottom',
    },
    headShake: {
      animationTimingFunction: 'ease-in-out',
    },
    swing: {
      transformOrigin: 'top center',
    },
    heartBeat: {
      animationDuration: processValue(heartBeatSpeed),
      animationTimingFunction: 'ease-in-out',
    },
    hinge: {
      animationDuration: processValue(hingeSpeed),
    },
    flip: {
      backfaceVisibility: 'visible',
    },
    flipInX: {
      backfaceVisibility: 'visible',
    },
    flipInY: {
      backfaceVisibility: 'visible',
    },
    flipOutX: {
      backfaceVisibility: 'visible',
    },
    flipOutY: {
      backfaceVisibility: 'visible',
    },
    bounceIn: {
      animationDuration: processValue(bounceInSpeed),
    },
    bounceOut: {
      animationDuration: processValue(bounceOutSpeed),
    },
  }

  return function ({ addUtilities, matchUtilities, theme }) {
    let animatedSpeed = options.animatedSpeed
      ? options.animatedSpeed
      : theme('animationDuration.DEFAULT')

    if (!animatedSpeed) animatedSpeed = 1000

    function processAnimate(value: any) {
      const framesName = `@keyframes ${camelCase(value)}`
      if (!availableKeyframes[framesName]) {
        console.warn(
          `invalid animation name: ${value}; arbitrary values not supported`
        )
      }

      let extras = {}
      if (Object.keys(extraProperties).includes(camelCase(value))) {
        extras = extraProperties[camelCase(value)]
      }

      return [
        { animationName: value, animationDuration: animatedSpeed, ...extras },
        { [`@keyframes ${value}`]: availableKeyframes[framesName] },
      ]
    }

    function filterTheme(name: string) {
      const values = theme(name)
      return Object.fromEntries(
        // @ts-expect-error values causes a type error
        Object.entries(values).filter(([key]) => key !== 'DEFAULT')
      )
    }

    matchUtilities(
      // @ts-expect-error typing does not allow for array of results
      { [`${PREFIX.replace(/-$/, '')}`]: (value) => processAnimate(value) },
      { values: createAnimateValues() }
    )

    matchUtilities(
      {
        [`${PREFIX}duration`]: (value) => ({
          animationDuration: processValue(value),
        }),
      },
      { values: filterTheme('animationDuration') }
    )

    matchUtilities(
      {
        [`${PREFIX}delay`]: (value) => ({
          animationDelay: processValue(value),
        }),
      },
      { values: filterTheme('animationDelay') }
    )

    matchUtilities(
      {
        [`${PREFIX}ease`]: (value) => ({
          animationTimingFunction: value as string,
        }),
      },
      { values: filterTheme('animationTimingFunction') }
    )

    addUtilities({
      [`.${PREFIX}running`]: { animationPlayState: 'running' },
      [`.${PREFIX}paused`]: { animationPlayState: 'paused' },
      [`.${PREFIX}infinite`]: { animationIterationCount: 'infinite' },
      [`.${PREFIX}delay`]: {
        animationDelay: processValue(animationDelaySpeed),
      },
    })

    matchUtilities(
      {
        [`${PREFIX}fill-mode`]: (value) => ({
          animationFillMode: value as string,
        }),
      },
      { values: filterTheme('animationFillMode') }
    )

    matchUtilities(
      {
        [`${PREFIX}direction`]: (value) => ({
          animationDirection: value as string,
        }),
      },
      { values: filterTheme('animationDirection') }
    )

    matchUtilities(
      {
        [`${PREFIX}repeat`]: (value) => ({
          animationIterationCount: value as string,
        }),
      },
      { values: filterTheme('animationRepeat') }
    )
  }
}

export default plugin.withOptions<AnimateOptions>(
  (options) => {
    return createPluginAnimate(options)
  },
  () => {
    return {
      content: {},
      theme: {
        extend: {
          animationDelay: ({ theme }) => ({
            1: '1s',
            2: '2s',
            3: '3s',
            4: '4s',
            5: '5s',
            2000: '2000ms',
            3000: '3000ms',
            4000: '4000ms',
            5000: '5000ms',
            ...theme('transitionDelay'),
          }),
          animationDuration: ({ theme }) => ({
            0: '0ms',
            1: '1s',
            2: '2s',
            3: '3s',
            4: '4s',
            5: '5s',
            2000: '2000ms',
            3000: '3000ms',
            4000: '4000ms',
            5000: '5000ms',
            ...theme('transitionDuration'),
            DEFAULT: '1000ms',
          }),
          animationTimingFunction: ({ theme }) => ({
            ...theme('transitionTimingFunction'),
          }),
          animationFillMode: {
            none: 'none',
            forwards: 'forwards',
            backwards: 'backwards',
            both: 'both',
          },
          animationDirection: {
            normal: 'normal',
            reverse: 'reverse',
            alternate: 'alternate',
            'alternate-reverse': 'alternate-reverse',
          },
          animationOpacity: ({ theme }) => ({
            ...theme('opacity'),
            DEFAULT: 0,
          }),
          animationTranslate: ({ theme }) => ({
            ...theme('translate'),
            DEFAULT: '100%',
          }),
          animationScale: ({ theme }) => ({
            ...theme('scale'),
            DEFAULT: 0,
          }),
          animationRotate: ({ theme }) => ({
            ...theme('rotate'),
            DEFAULT: '30deg',
          }),
          animationRepeat: {
            0: '0',
            1: '1',
            infinite: 'infinite',
          },
        },
      },
    } as any
  }
)
